Exclude dependencies of `std` for diagnostics

tgross35 opened the pull request:
Currently crates in the sysroot can show up in diagnostic suggestions, such as in https://github.com/rust-lang/rust/issues/135232. To prevent this, duplicate `all_traits` into `visible_traits` which only shows traits in non-private crates.

Setting `#![feature(rustc_private)]` overrides this and makes items in private crates visible as well, since `rustc_private` enables use of `std`'s private dependencies.

This may be reviewed per-commit.

Fixes: https://github.com/rust-lang/rust/issues/135232

tgross35 committed:
rustc_metadata: Rename `root` to `dep_root`

Currently `root` or `crate_root` is used to refer to an instance of
`CrateRoot` (representation of a crate's serialized metadata), but the
name `root` sometimes also refers to a `CratePath` representing a "root"
node in the dependency graph. In order to disambiguate, rename all
instances of the latter to `dep_root`.

tgross35 commented:
Unsure whether adding a query is okay or if this should just be a filtering method on `tcx`. I am still figuring out a good way to test this so

@rustbot author

tgross35 commented:
Another option is to drop `stdlib_private_dep` and just use `private_dep` in the query. I didn't do this originally because I thought `private_dep` was more tightly coupled to the `public_private_dependencies` feature, but maybe this isn't the case.

Open to suggestions here.

tgross35 committed:
Add a UI test for stdlib-private dependencies

Introduce a test that shows stdlib-private dependencies leaking into
diagnostics. This is resolved by a later commit.

tgross35 commented:
Okay, I think I have something workable here. I dropped `stdlib_private_dep` in favor of reusing `is_private_dep`, and added a repro.

@rustbot ready

SparrowLii reviewed (commented):
This change makes sense, but I'm not sure it is the best for diagnostics' design.
@compiler-errors Could you help have a review?

compiler-errors commented on compiler/rustc_metadata/src/rmeta/decoder/cstore_impl.rs:
```diff
@@ -561,6 +561,23 @@ pub(in crate::rmeta) fn provide(providers: &mut Providers) {
+                tcx.arena.alloc_from_iter(
+                    CStore::from_tcx(tcx)
+                        .iter_crate_data()
+                        .filter_map(|(cnum, meta)| (!meta.private_dep).then_some(cnum)),
```
can u just make this a `filter` then `map`? i think `(!bool).then_some` is so unreadable lol

compiler-errors replied:
like, i spent a few minutes going "why is this `private_dep.then_some()` before seeing that it was `(!private_dep).then_some()`

tgross35 replied:
Query is all gone

compiler-errors commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

bors commented:
:hourglass: Trying commit a04ed092c1918148495c4a40cd347a37286a916b with merge dbe10d00231af9fee521d76740718a1c644d81b1...
<!-- homu: {"type":"TryBuildStarted","head_sha":"a04ed092c1918148495c4a40cd347a37286a916b","merge_sha":"dbe10d00231af9fee521d76740718a1c644d81b1"} -->

compiler-errors commented on compiler/rustc_middle/src/ty/context.rs:
```diff
@@ -2078,9 +2078,10 @@ impl<'tcx> TyCtxt<'tcx> {
     pub fn all_traits(self) -> impl Iterator<Item = DefId> + 'tcx {
         iter::once(LOCAL_CRATE)
-            .chain(self.crates(()).iter().copied())
+            .chain(self.visible_crates(()).iter().copied())
```
yeah, this seems wrong. `all_traits` is also used by things like SMIR and rustdoc in ways that are not just diagnostics. if you want to distinguish "all visible traits" from "all traits", and then use the former in diagnostics, then it should be more explicit.

tgross35 replied:
I wasn't sure about this one, but yeah that makes sense. I guess there are a couple questions - should the query be kept or should it just filter as-needed? My original thought was that there could be other existing or new places where `crates` is used but `visible_crates` would be more correct, but maybe this isn't very true.

Then either way, should `all_traits` be split into `visible_traits`, or is filtering in `all_diagnostic_items` sufficient to cover this case? (I didn't realize that would affect the output here, so didn't actually test that on its own)

compiler-errors replied:
I think the filtering should be done as needed. if we expose `visible_traits` as just a function, it doesn't matter if the filtering is expensive b/c we only call it on the diagnostic codepath (i.e. failed compilation).

compiler-errors replied:
> Then either way, should all_traits be split into visible_traits, or is filtering in all_diagnostic_items sufficient to cover this case? (I didn't realize that would affect the output here, so didn't actually test that on its own)

probably worth just testing yourself, idk off the top of my head lol

tgross35 replied:
Just tested, seems like `all_diagnostic_items` doesn't have any effect on the relevant output here so it has to filter somewhere else too.

So I got rid of the query and split `all_traits` into `all_traits` and `visible_traits` that does the filtering on the fly. I also replaced `all_traits` in a few other places that deal with errors that seem reasonable (?), still waiting for my tests to finish and see if that affects much.

tgross35 committed:
Add `tcx.visible_traits()` and use it for producing diagnostics

Add an alternative to `tcx.all_traits()` that only shows traits that the
user might be able to use, for diagnostic purposes. With this available,
make use of it for diagnostics including associated type errors, which
is part of the problem with [1].

Includes a few comment updates for related API.

[1]: https://github.com/rust-lang/rust/issues/135232

tgross35 committed:
Make `#[rustc_private]` override dependency visibility

Really this is always-visible override only needs to happen when the
crate is a dependency of itself. However, this is a very internal
feature, so it doesn't seem worth doing any additional filtering here.

tgross35 commented on compiler/rustc_trait_selection/src/error_reporting/traits/fulfillment_errors.rs:
```diff
@@ -1730,7 +1730,7 @@ impl<'a, 'tcx> TypeErrCtxt<'a, 'tcx> {
         let trait_name = self.tcx.item_name(trait_def_id);
         let crate_name = self.tcx.crate_name(trait_def_id.krate);
-        if let Some(other_trait_def_id) = self.tcx.all_traits().find(|def_id| {
+        if let Some(other_trait_def_id) = self.tcx.visible_traits().find(|def_id| {
```
Ah I should drop this one, it's already checking the crate name

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/12741198681/job/35507467032) [(plain)](https://github.com/rust-lang/rust/commit/c0eb4afed10604db96487184fd3e8d220e695e64/checks/35507467032/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

```

</details>

tgross35 commented:
(^ spurious CI failure)

SparrowLii commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

bors commented:
:hourglass: Trying commit c0eb4afed10604db96487184fd3e8d220e695e64 with merge c9c7ddb9438f1b639c2379d6331bab066e152e92...
<!-- homu: {"type":"TryBuildStarted","head_sha":"c0eb4afed10604db96487184fd3e8d220e695e64","merge_sha":"c9c7ddb9438f1b639c2379d6331bab066e152e92"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/12742277418/job/35514313853)
Build commit: c9c7ddb9438f1b639c2379d6331bab066e152e92 (`c9c7ddb9438f1b639c2379d6331bab066e152e92`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/12742277418/job/35514313853"},"merge_sha":"c9c7ddb9438f1b639c2379d6331bab066e152e92"} -->

rust-timer commented:
Queued c9c7ddb9438f1b639c2379d6331bab066e152e92 with parent 047bc17d4ff3b3546187f7f6169eb059e8012b8c, future [comparison URL](https://perf.rust-lang.org/compare.html?start=047bc17d4ff3b3546187f7f6169eb059e8012b8c&end=c9c7ddb9438f1b639c2379d6331bab066e152e92).
There are currently 0 preceding artifacts in the [queue](https://perf.rust-lang.org/status.html).
It will probably take at least ~1.2 hours until the benchmark run finishes.
<!-- rust-timer: temporary -->

rust-timer commented:
Finished benchmarking commit (c9c7ddb9438f1b639c2379d6331bab066e152e92): [comparison URL](https://perf.rust-lang.org/compare.html?start=047bc17d4ff3b3546187f7f6169eb059e8012b8c&end=c9c7ddb9438f1b639c2379d6331bab066e152e92&stat=instructions:u).

### Overall result: no relevant changes - no action needed


Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

@bors rollup=never
@rustbot label: -S-waiting-on-perf -perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=047bc17d4ff3b3546187f7f6169eb059e8012b8c&end=c9c7ddb9438f1b639c2379d6331bab066e152e92&stat=instructions:u)
This benchmark run did not return any relevant results for this metric.

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=047bc17d4ff3b3546187f7f6169eb059e8012b8c&end=c9c7ddb9438f1b639c2379d6331bab066e152e92&stat=max-rss)
<details>
<summary>Results (primary -2.6%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -2.6% | [-2.9%, -2.3%] | 2     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | -2.6% | [-2.9%, -2.3%] | 2     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=047bc17d4ff3b3546187f7f6169eb059e8012b8c&end=c9c7ddb9438f1b639c2379d6331bab066e152e92&stat=cycles:u)
<details>
<summary>Results (secondary 1.0%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | 3.7%  | [3.0%, 4.5%]   | 3     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -3.2% | [-3.6%, -2.7%] | 2     |
| All ❌✅ (primary)                 | -     | -              | 0     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=047bc17d4ff3b3546187f7f6169eb059e8012b8c&end=c9c7ddb9438f1b639c2379d6331bab066e152e92&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 763.719s -> 763.091s (-0.08%)
**Artifact size**: 326.09 MiB -> 326.09 MiB (-0.00%)

SparrowLii commented on compiler/rustc_metadata/src/creader.rs:
```diff
@@ -402,7 +404,7 @@ impl<'a, 'tcx> CrateLoader<'a, 'tcx> {
         &mut self,
         host_lib: Option<Library>,
-        root: Option<&CratePaths>,
+        dep_root: Option<&CratePaths>,
```
I don't see the point of replacing `root` with `def_root` here, it might be simpler to keep the previous one

tgross35 replied:
The changes to this file are somewhat unrelated now, they were more relevant for the initial implementation. But I still think this makes sense, even if off topic - before this change, `root` could sometimes refer to a `CrateRoot` and sometimes to the dependency root's `CratePaths` (like here), it was pretty confusing to track through the call stack because the meaning switched back and forth. So this patch just makes the same name always refer to the same thing.

SparrowLii replied:
That's OK

SparrowLii reviewed (approved):
LGTM, just one nit

tgross35 commented:
@bors r=SparrowLii

bors commented:
:pushpin: Commit c0eb4afed10604db96487184fd3e8d220e695e64 has been approved by `SparrowLii`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=SparrowLii c0eb4afed10604db96487184fd3e8d220e695e64 -->
<!-- homu: {"type":"Approved","sha":"c0eb4afed10604db96487184fd3e8d220e695e64","approver":"SparrowLii","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

tgross35 commented:
@bors r-

Looks like this only fixes part of the problem :( https://github.com/rust-lang/rust/pull/135180#issuecomment-2588833249

tgross35 commented:
I don't get it. The error in that failure is coming from https://github.com/rust-lang/rust/blob/c0eb4afed10604db96487184fd3e8d220e695e64/compiler/rustc_hir_analysis/src/hir_ty_lowering/errors.rs#L211-L218, which after this change should only be looking at `visible_traits()`. Is `compiler_builtins` not actually marked private for some reason? More digging to do...

tgross35 committed:
Mark dependencies of the standard library as private by default

In order to avoid diagnostics suggesting stdlib-private dependencies,
make everything that is a direct dependency of any `std` crates private
by default. Note that this will be overridden, if the same crate is
public elsewhere in the crate graph then that overrides the private
default.

It may also be feasible to do this in the library crate, marking `std`'s
dependencies private via Cargo. However, given that the feature is still
rather unstable, doing this within the compiler seems more
straightforward.

Fixes: https://github.com/rust-lang/rust/issues/135232 [1]

tgross35 commented:
Alright, figured it out. `addr2line`, which I was using for my tests, was apparently already a private dependency. I am not sure why this is (possibly related to it being optional in `std`?) but just omitting private dependencies was enough to fix this. Other dependencies in the sysroot aren't marked private though, so could still sneak through. For my future reference, the list of available traits after filtering private crates, using the test that failed https://github.com/rust-lang/rust/pull/135180#issuecomment-2588833249:

```

     DefId(3:1748 ~ compiler_builtins[9bd7]::math::libm::support::float_traits::Float),
     DefId(3:1810 ~ compiler_builtins[9bd7]::math::libm::support::int_traits::MinInt),
     DefId(3:1821 ~ compiler_builtins[9bd7]::math::libm::support::int_traits::Int),
     DefId(3:1844 ~ compiler_builtins[9bd7]::math::libm::support::int_traits::DInt),
     DefId(3:1850 ~ compiler_builtins[9bd7]::math::libm::support::int_traits::HInt),
     DefId(3:1859 ~ compiler_builtins[9bd7]::math::libm::support::int_traits::CastInto),
     DefId(3:1862 ~ compiler_builtins[9bd7]::math::libm::support::int_traits::CastFrom),
     DefId(16:61 ~ gimli[25e7]::endianity::Endianity),
     DefId(16:135 ~ gimli[25e7]::read::util::sealed::Sealed),
     DefId(16:140 ~ gimli[25e7]::read::util::ArrayLike),
     DefId(16:354 ~ gimli[25e7]::read::cfi::UnwindOffset),
     DefId(16:363 ~ gimli[25e7]::read::cfi::_UnwindSectionPrivate),
     DefId(16:373 ~ gimli[25e7]::read::cfi::UnwindSection),
     DefId(16:512 ~ gimli[25e7]::read::cfi::UnwindContextStorage),
     DefId(16:1017 ~ gimli[25e7]::read::reader::ReaderOffset),
     DefId(16:1061 ~ gimli[25e7]::read::reader::Reader),
     DefId(16:1129 ~ gimli[25e7]::read::relocate::Relocate),
     DefId(16:1450 ~ gimli[25e7]::read::line::LineProgram),
     DefId(16:1745 ~ gimli[25e7]::read::lookup::LookupParser),
     DefId(16:1764 ~ gimli[25e7]::read::lookup::PubStuffEntry),
     DefId(16:1815 ~ gimli[25e7]::read::op::EvaluationStorage),
     DefId(16:2475 ~ gimli[25e7]::read::Section),
     DefId(18:140 ~ memchr[1cd9]::arch::all::packedpair::HeuristicFrequencyRank),
```

So at least `gimli`, `memchr` and `compiler_builtins` are public. I updated the test to use `gimli` instead and it did fail even with my fix, so I added a new commit that makes dependencies of std crates private by default.

SparrowLii commented on compiler/rustc_metadata/src/creader.rs:
```diff
@@ -390,15 +390,52 @@ impl<'a, 'tcx> CrateLoader<'a, 'tcx> {
+            // Not passed via extern and no metadata available. Not sure how we got here, but
+            // default to public.
+            (None, None) => false,
+        }
```
I guess something like this looks better
```
if extern_private.is_none() && dep_root.is_some_and(|d| STDLIB_STABLE_CRATES.contains(&d.name))
{
    return true;
}

match (extern_private, private_dep) {
    // Explicit non-private or default to public
    (Some(false), _) | (_, Some(false)) | (None, None) => false,
    // Explicit private
    _ => true,
}
```

SparrowLii commented:
I think the new CI [failure](https://github.com/rust-lang/rust/pull/135180#issuecomment-2589314871) unrelated with this change?

tgross35 commented:
The test I added here is the one that failed in that PR. I am assuming this is because Windows handles the sysroot differently somehow, since it passed in CI and for me locally. In the most recent push I marked the test `unix-only`

SparrowLii commented:
LGTM r=me if no other concerns

tgross35 commented:
That try job got further than it has before so

@bors r=SparrowLii

Thanks for reviewing

bors commented:
:pushpin: Commit ed63539282b7c087bfd6ef5e41584d1cf61f13f6 has been approved by `SparrowLii`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=SparrowLii ed63539282b7c087bfd6ef5e41584d1cf61f13f6 -->
<!-- homu: {"type":"Approved","sha":"ed63539282b7c087bfd6ef5e41584d1cf61f13f6","approver":"SparrowLii","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit ed63539282b7c087bfd6ef5e41584d1cf61f13f6 with merge 8c39ce5b4fb5b61796e5fd8cec56c7b9abd2122b...
<!-- homu: {"type":"BuildStarted","head_sha":"ed63539282b7c087bfd6ef5e41584d1cf61f13f6","merge_sha":"8c39ce5b4fb5b61796e5fd8cec56c7b9abd2122b"} -->

bjorn3 commented on compiler/rustc_middle/src/ty/util.rs:
```diff
@@ -876,6 +876,11 @@ impl<'tcx> TyCtxt<'tcx> {
     /// [direct]: rustc_session::cstore::ExternCrate::is_direct
     pub fn is_user_visible_dep(self, key: CrateNum) -> bool {
+        // `#![rustc_private]` overrides defaults to make private dependencies usable.
+        if self.features().enabled(sym::rustc_private) {
```
> However, this is a very internal feature, so it doesn't seem worth doing any additional filtering here.

Not really. There are a lot of external custom drivers that need to enable this feature.

tgross35 replied:
Are you suggesting that it is worth filtering here to only expose crates that are dependencies of `std`? Or just that the message is inaccurate.

bjorn3 replied:
I think the override for rustc_private should be removed. Even with rustc_private enabled you generally shouldn't do something like `extern crate libc;` from the sysroot.

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/12769514776/job/35602442066)
Approved by: SparrowLii
Pushing 8c39ce5b4fb5b61796e5fd8cec56c7b9abd2122b to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"SparrowLii","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/12769514776/job/35602442066"},"merge_sha":"8c39ce5b4fb5b61796e5fd8cec56c7b9abd2122b"} -->

tgross35 commented:
> #111076 marked all dependencies of libstd as private except for those that are actually part of the public interface of libstd to fix a similar diagnostic issue. This wasn't done for liballoc and other sysroot crates however. Should probably be fixed.

I was not aware of this history but that makes sense, I will update the others. In that case, do you think it makes sense to also revert ed63539282b7c087bfd6ef5e41584d1cf61f13f6? Assuming yes, since it would no longer be needed.

bjorn3 commented:
If the Cargo.toml changes are enough, then yes. It would make sense to me to revert that commit.

tgross35 commented:
Opened https://github.com/rust-lang/rust/pull/135501 as an experiment, but I have been unsuccessful keeping the test fixed just by setting `public = false`.

rust-timer commented:
Finished benchmarking commit (8c39ce5b4fb5b61796e5fd8cec56c7b9abd2122b): [comparison URL](https://perf.rust-lang.org/compare.html?start=c074d8eee765cfd64e6e143d2894c85c7f3ddc1d&end=8c39ce5b4fb5b61796e5fd8cec56c7b9abd2122b&stat=instructions:u).

### Overall result: no relevant changes - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=c074d8eee765cfd64e6e143d2894c85c7f3ddc1d&end=8c39ce5b4fb5b61796e5fd8cec56c7b9abd2122b&stat=instructions:u)
This benchmark run did not return any relevant results for this metric.

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=c074d8eee765cfd64e6e143d2894c85c7f3ddc1d&end=8c39ce5b4fb5b61796e5fd8cec56c7b9abd2122b&stat=max-rss)
<details>
<summary>Results (primary -1.6%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -1.6% | [-3.1%, -0.8%] | 4     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | -1.6% | [-3.1%, -0.8%] | 4     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=c074d8eee765cfd64e6e143d2894c85c7f3ddc1d&end=8c39ce5b4fb5b61796e5fd8cec56c7b9abd2122b&stat=cycles:u)
<details>
<summary>Results (primary 2.6%, secondary -3.4%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 2.6%  | [2.3%, 3.0%]   | 2     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -3.4% | [-4.6%, -2.6%] | 6     |
| All ❌✅ (primary)                 | 2.6%  | [2.3%, 3.0%]   | 2     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=c074d8eee765cfd64e6e143d2894c85c7f3ddc1d&end=8c39ce5b4fb5b61796e5fd8cec56c7b9abd2122b&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 763.379s -> 762.502s (-0.11%)
**Artifact size**: 326.08 MiB -> 326.12 MiB (0.01%)
